### ISSUE TITLE
Expression.Convert causing signature or security transparency is not compatible exception

### ISSUE THREAD

**Issue by Pondidum** [dadhi/FastExpressionCompiler#76]
Hi,

I am getting the following error when I use the FastExpressionCompiler:

```
ConversionTests.When_using_fast_expression_compilation

System.ArgumentException
Cannot bind to the target method because its signature or security transparency is not compatible with that of the delegate type.
   at System.Delegate.CreateDelegateNoSecurityCheck(Type type, Object target, RuntimeMethodHandle method)
   at System.Reflection.Emit.DynamicMethod.CreateDelegate(Type delegateType, Object target)
   at FastExpressionCompiler.ExpressionCompiler.TryCompile(ClosureInfo& closureInfo, Type delegateType, Type[] paramTypes, Type returnType, Object exprObj, ExpressionType exprNodeType, Type exprType, Object[] paramExprs, Boolean isNestedLambda)
   at FastExpressionCompiler.ExpressionCompiler.TryCompile[TDelegate](Expression bodyExpr, IList`1 paramExprs, Type[] paramTypes, Type returnType)
   at FastExpressionCompiler.ExpressionCompiler.TryCompile[TDelegate](LambdaExpression lambdaExpr)
   at FastExpressionCompiler.ExpressionCompiler.Compile[TDelegate](LambdaExpression lambdaExpr)

```

The following tests show as small a repro I can make - it works when using inbuilt expression compilation (`When_using_expression_compilation`) and fails on FastExpressionCompiler (`When_using_fast_expression_compilation`)

If I change the `TestTarget.ID` type to be a `Guid` and remove the  `var cast = Expression.Convert(idParam, typeof(CustomID));` line from `CreateWriter`, everything works fine.

Not sure if this is a bug or if I am doing something wrong, as this is the first time I am playing around with expression compilation.

```csharp

public class ConversionTests
{
    private Guid _id;
    private TestTarget _instance;

    public ConversionTests()
    {
        _id = Guid.NewGuid();
        _instance = new TestTarget();
    }

    [Fact]
    public void When_using_expression_compilation()
    {
        var expression = CreateWriter<TestTarget>();
        var write = expression.Compile();

        write(_instance, _id);

        _instance.ID.ShouldBe(new CustomID(_id));
    }

    [Fact]
    public void When_using_fast_expression_compilation()
    {
        var expression = CreateWriter<TestTarget>();
        var write = ExpressionCompiler.Compile<Action<TestTarget, Guid>>(expression);

        write(_instance, _id);

        _instance.ID.ShouldBe(new CustomID(_id));
    }

    [Fact]
    public void CanConvert()
    {
        var idParam = Expression.Parameter(typeof(Guid), "id");
        var cast = Expression.Convert(idParam, typeof(CustomID));
        var lambda = Expression.Lambda<Func<Guid, CustomID>>(cast, idParam);
        var convert = ExpressionCompiler.Compile<Func<Guid, CustomID>>(lambda);

        var id = Guid.NewGuid();
        var account = convert(id);
        account.ShouldBe(new CustomID(id));
    }

    private Expression<Action<T, Guid>> CreateWriter<T>()
    {
        var idMember = FindIdMember(typeof(T));

        var docParam = Expression.Parameter(typeof(T), "doc");
        var idParam = Expression.Parameter(typeof(Guid), "id");

        var cast = Expression.Convert(idParam, typeof(CustomID));

        var member = Expression.PropertyOrField(docParam, idMember.Name);
        var assign = Expression.Assign(member, cast);

        var lambda = Expression.Lambda<Action<T, Guid>>(assign, docParam, idParam);

        return lambda;
    }

    public static MemberInfo FindIdMember(Type type)
    {
        return type.GetProperties().FirstOrDefault(x => x.Name == "ID");
    }

    private class TestTarget
    {
        public CustomID ID { get; set; }
    }

    public struct CustomID : IEquatable<CustomID>
    {
        private readonly Guid _id;

        public CustomID(Guid id)
        {
            _id = id;
        }

        public bool Equals(CustomID other) => _id.Equals(other._id);

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            return obj is CustomID && Equals((CustomID) obj);
        }

        public override int GetHashCode() => _id.GetHashCode();
        public override string ToString() => _id.ToString();

        public static bool operator ==(CustomID left, CustomID right) => left.Equals(right);
        public static bool operator !=(CustomID left, CustomID right) => !left.Equals(right);

        public static implicit operator CustomID(Guid id) => new CustomID(id);
    }
}
```

**Comment by dadhi** [comment]
Hi @Pondidum ,

Thanks for the repro tests! 
From the first look the usage seems fine. 
Guessing that the `struct` together with `implicit operator` in it causing this issue, not sure at the moment. Will check soon.

**Comment by dadhi** [comment]
@Pondidum, Quick question, what is your version of FEC and the platform ?

Asking cause I have added your tests [here](https://github.com/dadhi/FastExpressionCompiler/blob/master/test/FastExpressionCompiler.IssueTests/Issue76_Expression_Convert_causing_signature_or_security_transparency_is_not_compatible_exception.cs), slightly cleaned (using `.CompileFast()`) and converted to NUnit, and they're passing just fine.

**Comment by Pondidum** [comment]
I will check and let you know what platform etc. tomorrow when I get back to work.
The FEC version was the latest off NuGet, as of about 6 hours ago, again will confirm when I get to work again.

**Comment by Pondidum** [comment]
@dadhi The application is targeting `4.6.1` (old csproj format also), and the FEC version is `1.10.0`

**Comment by dadhi** [comment]
Ok, the difference is really in usage of `ExpressionCompiler.Compile<Action<TestTarget, Guid>>(expression)` instead of `expression.CompileFast()` - latter is working, former is not.

The return type in `CompileFast` is properly detected as `void`.

I will try to fix it for `ExpressionCompiler.Compile`, __BUT__ it is already marked for deprecation - check the api doc on it.  So, I won't be trying too hard.

Please use the `CompileFast()` - it will fix the issue.

**Comment by Pondidum** [comment]
Thanks for the quick responses!  I'll change my code to use `CompileFast()`

**Comment by dadhi** [comment]
The fix was simple :-) I have released v1.10.1 on NuGet.
